packages/aws-cdk/lib/api/hotswap/s3-bucket-deployments.ts

```diff
@@ -82,20 +101,47 @@ async function changeIsForS3DeployCustomResourcePolicy(
           // If S3Deployment -> Lambda -> Role and IAM::Policy -> Role, then this IAM::Policy change is an
           // artifact of old-style synthesis
           if (lambdaRef.Type !== 'Custom::CDKBucketDeployment') {
-            return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+            return [{
+              hotswappable: false,
+              reason: `found an AWS::IAM::Policy that has Role '${roleLogicalId}' that is referred to by AWS::Lambda::Function '${roleRef.LogicalId}' that is referred to by ${lambdaRef.Type} '${lambdaRef.LogicalId}', which does not have type 'Custom::CDKBucketDeployment'`,
+              rejectedChanges: [],
+              logicalId: iamPolicyLogicalId,
+              resourceType: change.newValue.Type,
+            }];
           }
         }
       } else if (roleRef.Type === 'AWS::IAM::Policy') {
         if (roleRef.LogicalId !== iamPolicyLogicalId) {
-          return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+          return [{
+            hotswappable: false,
+            reason: `found an AWS::IAM::Policy that has Role '${roleLogicalId}' that is referred to by AWS::IAM::Policy '${roleRef.LogicalId}' that is not the policy of the s3 bucket deployment`,
+            rejectedChanges: [],
+            logicalId: roleRef.LogicalId,
+            resourceType: change.newValue.Type,
+          }];
         }
       } else {
-        return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+        return [{
+          hotswappable: false,
+          reason: `found a reference to the role '${roleLogicalId}' that is not of type AWS::Lambda::Function or AWS::IAM::Policy, so the bucket deployment cannot be hotswapped`,
+          rejectedChanges: [],
+          logicalId: roleRef.LogicalId,
+          resourceType: change.newValue.Type,
+        }];
       }
     }
   }
 
-  return ChangeHotswapImpact.IRRELEVANT;
+  // this doesn't block the hotswap, but it also isn't a hotswappable change by itself. Return
+  // an empty change to signify this.
+  return [{
```

Comment:
If there's nothing to do, can't we just `return [];` ? Or is it important that this shows up in a report somewhere?

Reply:
If we return `[]` it will become a change that the detectors have no opinion on. That messes up the Classic mode deployments, which interpret that to be a change to a resource we don't allow hotswapping on instead of an ignored change.